src/java/org/apache/cassandra/utils/JVMStabilityInspector.java

```diff
@@ -125,7 +120,28 @@ public static void inspectThrowable(Throwable t, boolean propagateOutOfMemory, C
             killer.killCurrentJVM(t);
 
         if (t.getCause() != null)
-            inspectThrowable(t.getCause(), propagateOutOfMemory, fn);
+            inspectThrowable(t.getCause(), fn);
+    }
+
+    /**
+     * Intentionally produce a heap space OOM upon seeing a Direct buffer memory OOM.
+     * Direct buffer OOM cannot trigger JVM OOM error related options,
+     * e.g. OnOutOfMemoryError, HeapDumpOnOutOfMemoryError, etc.
+     */
+    private static void forceHeapSpaceOomMaybe(OutOfMemoryError oom)
+    {
+        // See the oom thrown from java.nio.Bits.reserveMemory.
+        if (oom.getMessage() != null && !oom.getMessage().equals("Direct buffer memory"))
+        {
+            return;
+        }
+        logger.warn("Force heap space OutOfMemoryError in the presence of", oom);
+        while (true)
+        {
+            // java.util.AbstractCollection.MAX_ARRAY_SIZE is defined as Integer.MAX_VALUE - 8
+            // so Integer.MAX_VALUE / 2 should be a large enough and safe size to request.
+            long[] ignored = new long[Integer.MAX_VALUE / 2];
```

Comment:
it may be good to collect the references in a list to make sure GC isn't able to free right away.

```
List<long[]> ignored = new ArrayList<>();
while (true)
{
    ignored.add(new long[Integer.MAX_VALUE / 2]);
}
```

Reply:
Also, in case the while true loop never breaks (such as JVM gets smart enough to know this is not needed and avoids the allocation), we should probably start logging an error saying that this trick isn't doing what we hoped.

Reply:
spoke in slack, adding `@Exclude` to the method should remove the concern of "smartness" trying to avoid the allocation